File: 114.flatten-binary-tree-to-linked-list.py
#
# @lc app=leetcode id=114 lang=python3
#
# [114] Flatten Binary Tree to Linked List
#


# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root:
            return None
        self.flattenHelper(root)

    def flattenHelper(self, root: TreeNode) -> TreeNode:
        # flatten and return the last node
        if not root.left and not root.right:
            return root
        if not root.left:
            return self.flattenHelper(root.right)
        if not root.right:
            leftTail = self.flattenHelper(root.left)
            root.right = root.left
            root.left = None
            return leftTail
        leftTail = self.flattenHelper(root.left)
        rightTail = self.flattenHelper(root.right)
        leftTail.right = root.right
        root.right = root.left
        root.left = None
        return rightTail


# @lc code=end

File: 114.flatten-binary-tree-to-linked-list.go
/*
 * @lc app=leetcode id=114 lang=golang
 *
 * [114] Flatten Binary Tree to Linked List
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
package main

/*
recursively call flatten
for each tree root,
1. call flatten on its left subtree
2. call flatten on its right subtree
4. append the right subtree (right linked list) behind the left subtree (if right tree is not nil)
3. move left linked list (left subtree) to right child
4. set left child to null

base cases
1. root is nil -> return nil
*/
func flatten(root *TreeNode) {
	if root == nil {
		return
	}
	flatten(root.Left)
	flatten(root.Right)
	if root.Right != nil {
		if root.Left == nil {
			root.Left = root.Right
		} else {
			// find the tail of left list
			curr := root.Left
			for curr != nil && curr.Right != nil {
				curr = curr.Right
			}
			curr.Right = root.Right
		}
	}
	root.Right = root.Left
	root.Left = nil
}

// @lc code=end
